### PR TITLE
feat(shorebird_cli): `shorebird preview` iOS support

### LINKED ISSUE #883: feat: `shorebird preview` should support iOS

### PR DESCRIPTION
**Pull request by felangel**
## Description

- feat(shorebird_cli): `shorebird preview` iOS support (closes #883)

## Type of Change

- [X] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: packages/shorebird_cli/lib/src/commands/preview_command.dart [hunk 1]
```diff
@@ -206,7 +209,42 @@ class PreviewCommand extends ShorebirdCommand
...
+    );
+
+    if (!Directory(runnerPath).existsSync()) {
+      final downloadArtifactProgress = logger.progress('Downloading release');
+      try {
+        final releaseRunnerArtifact =
+            await codePushClientWrapper.getReleaseArtifact(
+          appId: appId,
+          releaseId: release.id,
+          arch: 'runner',
```
Comment (eseidel) [inline packages/shorebird_cli/lib/src/commands/preview_command.dart:230]: Did you mean arch runner?

File: packages/shorebird_cli/lib/src/commands/preview_command.dart [hunk 2]
```diff
@@ -206,7 +209,42 @@ class PreviewCommand extends ShorebirdCommand
   }
 
   Future<int> installAndLaunchIos(String appId, Release release) async {
-    return ExitCode.unavailable.code;
+    const platform = ReleasePlatform.ios;
+    final previewDirectory = cache.getPreviewDirectory(appId);
+    final runnerPath = p.join(
+      previewDirectory.path,
+      '${platform.name}_${release.version}.app',
```
Comment (eseidel) [inline packages/shorebird_cli/lib/src/commands/preview_command.dart]: You repeat '${platform.name}_${release.version}' in this file, which is fine, but you might want a function 🤷

File: packages/shorebird_cli/lib/src/commands/preview_command.dart (hunk 1 above)
Comment (eseidel) [reply packages/shorebird_cli/lib/src/commands/preview_command.dart:230]: You'll want a named constant at least.

File: packages/shorebird_cli/lib/src/commands/preview_command.dart [hunk 3]
```diff
@@ -222,4 +260,24 @@ extension on String {
     final file = File(path)..createSync(recursive: true);
     await response.pipe(file.openWrite());
   }
+
+  Future<void> downloadAndExtract(String path) async {
+    final uri = Uri.parse(this);
+    final client = HttpClient();
```
Comment (eseidel) [inline packages/shorebird_cli/lib/src/commands/preview_command.dart:285]: How is this mockable?

File: packages/shorebird_cli/lib/src/commands/preview_command.dart [hunk 4]
```diff
@@ -222,4 +260,24 @@ extension on String {
...
+    final request = await client.getUrl(uri);
+    final response = await request.close();
+    if (response.statusCode != 200) {
+      throw Exception('Failed to download artifact at $this');
+    }
+    final tempDir = Directory.systemTemp.createTempSync();
```
Comment (eseidel) [inline packages/shorebird_cli/lib/src/commands/preview_command.dart:291]: You probably want someone to pass you a FileSystem too?

Comment (eseidel) [inline packages/shorebird_cli/lib/src/commands/preview_command.dart:291]: You probably want someone to pass you a FileSystem too?

File: packages/shorebird_cli/lib/src/commands/preview_command.dart (hunk 1 above)
Comment (felangel) [reply packages/shorebird_cli/lib/src/commands/preview_command.dart:230]: Yeah this is the naming convention. I think we can create some constants to clean this up in a subsequent PR.

File: packages/shorebird_cli/lib/src/commands/preview_command.dart (hunk 2 above)
Comment (felangel) [reply packages/shorebird_cli/lib/src/commands/preview_command.dart]: Where is this repeated? Did you mean one in iOS and one in Android?

File: packages/shorebird_cli/lib/src/commands/preview_command.dart [hunk 5]
```diff
@@ -128,7 +131,7 @@ class PreviewCommand extends ShorebirdCommand
     final previewDirectory = cache.getPreviewDirectory(appId);
     final aabPath = p.join(
       previewDirectory.path,
-      '${platform}_${release.version}.aab',
+      '${platform.name}_${release.version}.aab',
```
Comment (eseidel) [inline packages/shorebird_cli/lib/src/commands/preview_command.dart]: here

File: packages/shorebird_cli/lib/src/commands/preview_command.dart [hunk 6]
```diff
@@ -162,7 +165,7 @@ class PreviewCommand extends ShorebirdCommand
 
     final apksPath = p.join(
       previewDirectory.path,
-      '${platform}_${release.version}.apks',
+      '${platform.name}_${release.version}.apks',
```
Comment (eseidel) [inline packages/shorebird_cli/lib/src/commands/preview_command.dart]: here

File: packages/shorebird_cli/lib/src/commands/preview_command.dart (hunk 3 above)
Comment (felangel) [reply packages/shorebird_cli/lib/src/commands/preview_command.dart:285]: Via `HttpOverriddes.runZoned`

File: packages/shorebird_cli/lib/src/ios_deploy.dart [hunk 7]
```diff
@@ -24,25 +44,265 @@ class IOSDeploy {
...
+
+        previousLine = line;
+      }
+
+      void onStderr(String line) {
+        monitorIOSDeployFailure(line, logger);
+        logger.detail(line);
+      }
+
+      final stdoutSubscription = launchProcess.stdout
```
Comment (eseidel) [inline packages/shorebird_cli/lib/src/ios_deploy.dart]: subscribeToLine(launchProcess.stdout, onStdOut) or similar?  Not important, just a little surprised you had to write this yourself.

File: packages/shorebird_cli/lib/src/ios_deploy.dart [hunk 8]
```diff
@@ -24,25 +44,265 @@ class IOSDeploy {
...
+
+      final stderrSubscription = launchProcess.stderr
+          .transform<String>(utf8.decoder)
+          .transform<String>(const LineSplitter())
+          .listen(onStderr);
+
+      final status = await launchProcess.exitCode;
+      logger.detail('ios-deploy exited with code $exitCode');
+      debuggerState = _IOSDeployDebuggerState.detached;
+      unawaited(stdoutSubscription.cancel());
```
Comment (eseidel) [inline packages/shorebird_cli/lib/src/ios_deploy.dart:285]: And this.  I guess this is Dart missing RAIIAs.

File: packages/shorebird_cli/lib/src/ios_deploy.dart (hunk 7 above)
Comment (felangel) [reply packages/shorebird_cli/lib/src/ios_deploy.dart]: Yeah good call let me create a helper for this to at least simplify things a bit more.

File: packages/shorebird_cli/lib/src/commands/preview_command.dart (hunk 4 above)
Comment (felangel) [reply packages/shorebird_cli/lib/src/commands/preview_command.dart:291]: Just staying consistent with other code but we can refactor to use a FS wrapper in future PRs.

File: packages/shorebird_cli/lib/src/commands/preview_command.dart [hunk 9]
```diff
@@ -222,4 +277,24 @@ extension on String {
     final file = File(path)..createSync(recursive: true);
     await response.pipe(file.openWrite());
   }
+
+  Future<void> downloadAndExtract(String path) async {
```
Comment (bryanoltman) [inline packages/shorebird_cli/lib/src/commands/preview_command.dart]: Nit: named parameter would be useful here (eg. `downloadAndExtract({required String outputPath})` or similar)

**Review by bryanoltman — APPROVED** [review]

File: packages/shorebird_cli/lib/src/commands/preview_command.dart (hunk 9 above)
Comment (felangel) [reply packages/shorebird_cli/lib/src/commands/preview_command.dart]: Good call, done 👍

**Pull request merged**